File: Java/Szyfr Gronsfelda.java
import java.util.Scanner;

public class Main
{
    public static String gronsfeld(String text, String k, boolean encrypt) 
    {
        StringBuilder result = new StringBuilder();
        int keyLength = k.length();
        for (int i = 0; i < text.length(); i++) 
        {
            char letter = text.charAt(i);
            int shift = Character.getNumericValue(k.charAt(i % keyLength));
            if (!encrypt)
                shift = -shift;
            char newLetter = (char) ((letter - 'A' + shift + 26) % 26 + 'A');
            result.append(newLetter);
        }
        return result.toString();
    }

    public static void main(String[] args) 
    {
        Scanner input = new Scanner(System.in);
        String mode = input.nextLine();
        String k = input.nextLine();
        String text = input.nextLine();
        if (mode.equals("SZYFRUJ"))
            System.out.println(gronsfeld(text, k, true));
        else
            System.out.println(gronsfeld(text, k, false));
        input.close();
    }
}

File: C++/Szyfr Gronsfelda.cpp
#include <string>
#include <iostream>

void gronsfeld(std::string& mode, std::string& k, std::string& text) 
{
    int keyLength = k.length();
    int n = text.length();
    for (int i = 0; i < n; i++) 
    {
        int shift = k[i % keyLength] - '0';
        if (mode == "SZYFRUJ")
            text[i] = 'A' + (text[i] - 'A' + shift) % 26;
        else
            text[i] = 'A' + (text[i] - 'A' - shift + 26) % 26;
    }
    std::cout << text << '\n';
}

int main() 
{
    std::string mode, k, text;
    std::cin >> mode >> k >> text;
    gronsfeld(mode, k, text);
    return 0;
}